tests/test_notifier.py

```diff
@@ -0,0 +1,341 @@
+                             DSS_NOTIFY_DELAYS="3 2 1 .5",
+                             DSS_NOTIFY_WORKERS="7"):
+            notifier = Notifier.from_config()
+            self.assertEqual([3.0, 2.0, 1.0, .5], notifier._delays)
+            self.assertEqual(7, notifier._num_workers)
+            self.assertTrue(Config.notification_is_async())
+            self.assertEqual(Config.notification_attempts(), 4)
+
+
+class TestWorkerQueueAssignment(unittest.TestCase):
+    def _test(self, num_workers, queue_lengths):
+        notifier = Notifier(deployment_stage='foo', delays=[0] * len(queue_lengths), num_workers=num_workers)
+        return list(notifier._work_queue_indices(queue_lengths))
+
+    def test_edge_cases(self):
+        self.assertEquals(self._test(num_workers=1, queue_lengths=[0]), [0])
+        self.assertEquals(self._test(num_workers=2, queue_lengths=[0]), [0, 0])
+        self.assertIn(self._test(num_workers=1, queue_lengths=[0, 0]), ([0], [1]))
+
+    def test_worker_surplus(self):
+        num_workers = 10
+        repeats = 1000
+        imbalance = .5
+        queue_coverage = Counter()
+        for i in range(repeats):
+            # N workers for two queues, one queue shorter than the other
+            queue_indices = self._test(num_workers=num_workers,
+                                       queue_lengths=[100, imbalance * 100])
+            # Every queue should be served by at least one worker (mandatory services)
+            self.assertEquals(set(queue_indices), {0, 1})
+            # Count how many times each queue was served
+            queue_coverage.update(queue_indices)
+        # Every worker in every iteration serves exactly one queue
+        self.assertEquals(sum(queue_coverage.values()), repeats * num_workers)
+        # Compute the probability of the longer queue being served by a surplus workers
+        service_ratio = (queue_coverage[1] - repeats) / (queue_coverage[0] - repeats)
+        # The probability should be equal to the imbalance, within an epsilon
+        self.assertAlmostEqual(service_ratio, imbalance, delta=.1)
+
+    def test_worker_shortage(self):
+        num_queues = 10
+        num_workers = 2
+        repeats = 1000
+        imbalance = .5
+        queue_coverage = Counter()
+        for i in range(repeats):
+            queue_indices = self._test(num_workers=num_workers,
+                                       queue_lengths=[100] + [round(imbalance * 100)] * (num_queues - 1))
+            queue_coverage.update(queue_indices)
+        # Every worker in every iteration serves exactly one queue
+        self.assertEquals(sum(queue_coverage.values()), repeats * num_workers)
+        first_queue_coverage = queue_coverage.pop(0)
+        avg = sum(c for c in queue_coverage.values()) / (num_queues - 1)
+        self.assertAlmostEqual(avg / first_queue_coverage, imbalance, delta=.1)
+        sigma = sqrt(sum((c - avg) ** 2 for c in queue_coverage.values()) / (num_queues - 2))
```

Comment:
it's not immediately clear what this is proving. Add some comments for clarity.

Reply:
Added comments.